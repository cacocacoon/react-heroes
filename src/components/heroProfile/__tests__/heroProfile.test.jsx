import React from 'react'
import thunk from 'redux-thunk'
import { MemoryRouter as Router } from 'react-router-dom'
import { cleanup, render, fireEvent } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import TestRenderer, { act } from 'react-test-renderer'

import HeroProfile from '../heroProfile'

global.fetch = jest.fn().mockImplementation(() => (
	Promise.resolve({
		json: () => Promise.resolve({})
	})
))

afterEach(cleanup)
const mockStore = configureMockStore([thunk])

function TestHeroProfile({ store, heroId }) {
	return (
		<Provider store={store} >
			<Router>
				<HeroProfile match={{ params: { heroId } }} />
			</Router>
		</Provider>
	)
}

const store = mockStore({
	hero: {
		profiles: { '1': {
			str: 2,
			int: 7,
			agi: 9,
			luk: 7,
			maxPoints: 25
		}}
	}
})
describe('heroProfile', () => {
	test('dont show', async () => {
		let heroProfile
		await act(async () => {
			heroProfile = TestRenderer.create(
				<TestHeroProfile store={store} />
			)
		})

		expect(heroProfile.toJSON()).toMatchSnapshot()
	})

	test('loading', async () => {
		let heroProfile
		await act(async () => {
			heroProfile = TestRenderer.create(
				<TestHeroProfile store={store} heroId="2" />
			)
		})

		expect(heroProfile.toJSON()).toMatchSnapshot()
	})

	test('show profile', async () => {
		let heroProfile
		await act(async () => {
			heroProfile = TestRenderer.create(
				<TestHeroProfile store={store} heroId="1" />
			)
		})

		expect(heroProfile.toJSON()).toMatchSnapshot()
	})

	test('save profile success', async () => {
		store.dispatch = jest.fn(store.dispatch)
		global.fetch = jest.fn().mockImplementationOnce(() => (
			Promise.resolve({
				text: () => Promise.resolve('ok')
			})
		))
		await act(async () => {
			render(<TestHeroProfile store={store} heroId="1" />)
		})

		const button = document.getElementsByClassName('save-button')[0]
		act(() => {
			fireEvent.click(button)
		})
		const response = await store.dispatch.mock.results[1].value
		expect(response).toBe('ok')
	})
})