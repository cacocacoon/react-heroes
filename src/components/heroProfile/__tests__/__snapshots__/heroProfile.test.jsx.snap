// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heroProfile dont show 1`] = `null`;

exports[`heroProfile loading 1`] = `
<div
  className="hero-profile"
>
  <div
    className="ant-divider ant-divider-horizontal ant-divider-dashed"
    role="separator"
  />
  <div
    className="ant-skeleton ant-skeleton-active"
  >
    <div
      className="ant-skeleton-content"
    >
      <h3
        className="ant-skeleton-title"
        style={
          Object {
            "width": "38%",
          }
        }
      />
      <ul
        className="ant-skeleton-paragraph"
      >
        <li
          style={
            Object {
              "width": undefined,
            }
          }
        />
        <li
          style={
            Object {
              "width": undefined,
            }
          }
        />
        <li
          style={
            Object {
              "width": undefined,
            }
          }
        />
        <li
          style={
            Object {
              "width": "61%",
            }
          }
        />
      </ul>
    </div>
  </div>
</div>
`;

exports[`heroProfile show profile 1`] = `
<div
  className="hero-profile"
>
  <div
    className="ant-divider ant-divider-horizontal ant-divider-dashed"
    role="separator"
  />
  <div
    className="ant-row-flex ant-row-flex-bottom"
  >
    <div
      className="ant-col ant-col-xs-24 ant-col-sm-12 ant-col-md-12"
    >
      <div
        className="ability"
      >
        <p>
          STR
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: plus"
            className="anticon anticon-plus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </i>
        </button>
        <p>
          2
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: minus"
            className="anticon anticon-minus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="minus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </i>
        </button>
      </div>
      <div
        className="ability"
      >
        <p>
          INT
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: plus"
            className="anticon anticon-plus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </i>
        </button>
        <p>
          7
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: minus"
            className="anticon anticon-minus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="minus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </i>
        </button>
      </div>
      <div
        className="ability"
      >
        <p>
          AGI
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: plus"
            className="anticon anticon-plus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </i>
        </button>
        <p>
          9
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: minus"
            className="anticon anticon-minus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="minus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </i>
        </button>
      </div>
      <div
        className="ability"
      >
        <p>
          LUK
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: plus"
            className="anticon anticon-plus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </i>
        </button>
        <p>
          7
        </p>
        <button
          className="ant-btn ant-btn-circle ant-btn-icon-only"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            aria-label="icon: minus"
            className="anticon anticon-minus"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="minus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </i>
        </button>
      </div>
    </div>
    <div
      className="ant-col ant-col-xs-24 ant-col-sm-12 ant-col-md-12"
    >
      <div
        className="right-panel"
      >
        <p
          className="remaining"
        >
          剩餘點數：
          0
        </p>
        <button
          className="ant-btn save-button ant-btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            儲 存
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
